README.md
ar_sys
======

3D pose estimation ROS package using ArUco marker boards

## Usage:
- roslaunch ar_sys single_board.launch
- roslaunch ar_sys multi_boards.launch
- roslaunch ar_sys system_viewer.launch


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package ar_sys
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Forthcoming
-----------
* Added arguments to launch files
* Contributors: Hamdi Sahloul

1.0.4 (2014-09-26)
------------------
* Added install instructions
* Contributors: Hamdi Sahloul

1.0.3 (2014-09-23)
------------------
* Changed size to support small screens
* Prefixed frame_ids by / for topic_tools/relay compatibility in groovy
* Employed topic_tools/relay to monitor both single and multi boards transform
* Update the package URL
* Remove false dependency
* Correct the default path of boards_config
* Contributors: Hamdi Sahloul

1.0.2 (2014-09-17)
------------------
* Update the package to declare its visualization_msgs dependency
* Initial tracks.yaml
* Contributors: Hamdi Sahloul

1.0.1 (2014-09-17)
------------------
* Fix the single marker coordinates to match ROS
* Introducing system viewer; a node that is capable of handling input from multi-cameras and display all the boards and virtual relative points in 3D using Rviz
* Replaced arrays and its count by vectors of enums
* Fixed the rotation of ArUco and OpenCV coordinates to match the camera and ROS specifications
* Enhanced time and source tracking of images
* Prevented a buffer overflow and made a naming convention
* Enhanced the multi detector performance by drawing the markers once, calculating the markers once using the minimum indicated size
* Fixed a bug in identification due to changes made to the orginal image
* Added launch command for the multi boards detector
* Added multi-boards detection capability
* Added a test board image for testing purposes
* Initial package: single board detection
* Contributors: Hamdi Sahloul


src/utils.cpp
/**
* @file utils.cpp
* @author Bence Magyar
* @date June 2012
* @version 0.1
* @brief ROS2ArUco utilities.
*/

#include <ar_sys/utils.h>
#include <ros/console.h>
#include <iostream>
#include <tf/transform_datatypes.h>
#include <opencv2/calib3d/calib3d.hpp>

aruco::CameraParameters ar_sys::getCamParams(const sensor_msgs::CameraInfo& cam_info,
                                                                bool useRectifiedParameters)
{
    cv::Mat cameraMatrix(3, 3, CV_32FC1);
    cv::Mat distorsionCoeff(4, 1, CV_32FC1);
    cv::Size size(cam_info.height, cam_info.width);

    if ( useRectifiedParameters )
    {
      cameraMatrix.setTo(0);
      cameraMatrix.at<float>(0,0) = cam_info.P[0];   cameraMatrix.at<float>(0,1) = cam_info.P[1];   cameraMatrix.at<float>(0,2) = cam_info.P[2];
      cameraMatrix.at<float>(1,0) = cam_info.P[4];   cameraMatrix.at<float>(1,1) = cam_info.P[5];   cameraMatrix.at<float>(1,2) = cam_info.P[6];
      cameraMatrix.at<float>(2,0) = cam_info.P[8];   cameraMatrix.at<float>(2,1) = cam_info.P[9];   cameraMatrix.at<float>(2,2) = cam_info.P[10];

      for(int i=0; i<4; ++i)
        distorsionCoeff.at<float>(i, 0) = 0;
    }
    else
    {
      for(int i=0; i<9; ++i)
        cameraMatrix.at<float>(i%3, i-(i%3)*3) = cam_info.K[i];

      for(int i=0; i<4; ++i)
        distorsionCoeff.at<float>(i, 0) = cam_info.D[i];
    }

    return aruco::CameraParameters(cameraMatrix, distorsionCoeff, size);
}

tf::Transform ar_sys::getTf(const cv::Mat &Rvec, const cv::Mat &Tvec)
{
    cv::Mat rot(3, 3, CV_32FC1);
    cv::Rodrigues(Rvec, rot);

    cv::Mat rotate_to_sys(3, 3, CV_32FC1);
    // -1 0 0
    // 0 0 1
    // 0 1 0
    rotate_to_sys.at<float>(0,0) = -1.0;
    rotate_to_sys.at<float>(0,1) = 0.0;
    rotate_to_sys.at<float>(0,2) = 0.0;
    rotate_to_sys.at<float>(1,0) = 0.0;
    rotate_to_sys.at<float>(1,1) = 0.0;
    rotate_to_sys.at<float>(1,2) = 1.0;
    rotate_to_sys.at<float>(2,0) = 0.0;
    rotate_to_sys.at<float>(2,1) = 1.0;
    rotate_to_sys.at<float>(2,2) = 0.0;
    rot = rot*rotate_to_sys.t();

    tf::Matrix3x3 tf_rot(rot.at<float>(0,0), rot.at<float>(0,1), rot.at<float>(0,2),
                       rot.at<float>(1,0), rot.at<float>(1,1), rot.at<float>(1,2),
                       rot.at<float>(2,0), rot.at<float>(2,1), rot.at<float>(2,2));

    tf::Vector3 tf_orig(Tvec.at<float>(0,0), Tvec.at<float>(1,0), Tvec.at<float>(2,0));

    return tf::Transform(tf_rot, tf_orig);
}


src/single_board.cpp
/**
* @file single_board.cpp
* @author Hamdi Sahloul
* @date September 2014
* @version 0.1
* @brief ROS version of the example named "simple_board" in the Aruco software package.
*/

#include <iostream>
#include <fstream>
#include <sstream>
#include <aruco/aruco.h>
#include <aruco/boarddetector.h>
#include <aruco/cvdrawingutils.h>

#include <opencv2/core/core.hpp>
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/image_encodings.h>
#include <ar_sys/utils.h>
#include <tf/transform_broadcaster.h>
#include <tf/transform_listener.h>

using namespace aruco;

class ArSysSingleBoard
{
	private:
		cv::Mat inImage, resultImg;
		aruco::CameraParameters camParam;
		bool useRectifiedImages;
		bool draw_markers;
		bool draw_markers_cube;
		bool draw_markers_axis;
		MarkerDetector mDetector;
		vector<Marker> markers;
		BoardConfiguration the_board_config;
		BoardDetector the_board_detector;
		Board the_board_detected;
		ros::Subscriber cam_info_sub;
		bool cam_info_received;
		image_transport::Publisher image_pub;
		image_transport::Publisher debug_pub;
		ros::Publisher pose_pub;
		ros::Publisher transform_pub; 
		ros::Publisher position_pub;
		std::string board_frame;

		double marker_size;
		std::string board_config;

		ros::NodeHandle nh;
		image_transport::ImageTransport it;
		image_transport::Subscriber image_sub;

		tf::TransformListener _tfListener;

	public:
		ArSysSingleBoard()
			: cam_info_received(false),
			nh("~"),
			it(nh)
		{
			image_sub = it.subscribe("/image", 1, &ArSysSingleBoard::image_callback, this);
			cam_info_sub = nh.subscribe("/camera_info", 1, &ArSysSingleBoard::cam_info_callback, this);

			image_pub = it.advertise("result", 1);
			debug_pub = it.advertise("debug", 1);
			pose_pub = nh.advertise<geometry_msgs::PoseStamped>("pose", 100);
			transform_pub = nh.advertise<geometry_msgs::TransformStamped>("transform", 100);
			position_pub = nh.advertise<geometry_msgs::Vector3Stamped>("position", 100);

			nh.param<double>("marker_size", marker_size, 0.05);
			nh.param<std::string>("board_config", board_config, "boardConfiguration.yml");
			nh.param<std::string>("board_frame", board_frame, "");
			nh.param<bool>("image_is_rectified", useRectifiedImages, true);
			nh.param<bool>("draw_markers", draw_markers, false);
			nh.param<bool>("draw_markers_cube", draw_markers_cube, false);
			nh.param<bool>("draw_markers_axis", draw_markers_axis, false);

			the_board_config.readFromFile(board_config.c_str());

			ROS_INFO("ArSys node started with marker size of %f m and board configuration: %s",
					 marker_size, board_config.c_str());
		}

		void image_callback(const sensor_msgs::ImageConstPtr& msg)
		{
			static tf::TransformBroadcaster br;
			if(!cam_info_received) return;

			cv_bridge::CvImagePtr cv_ptr;
			try
			{
				cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::RGB8);
				inImage = cv_ptr->image;
				resultImg = cv_ptr->image.clone();

				//detection results will go into "markers"
				markers.clear();
				//Ok, let's detect
				mDetector.detect(inImage, markers, camParam, marker_size, false);
				//Detection of the board
				float probDetect=the_board_detector.detect(markers, the_board_config, the_board_detected, camParam, marker_size);
				if (probDetect > 0.0)
				{
					tf::Transform transform = ar_sys::getTf(the_board_detected.Rvec, the_board_detected.Tvec);

					tf::StampedTransform stampedTransform(transform, msg->header.stamp, msg->header.frame_id, board_frame);
					br.sendTransform(stampedTransform);
					geometry_msgs::PoseStamped poseMsg;
					tf::poseTFToMsg(transform, poseMsg.pose);
					poseMsg.header.frame_id = msg->header.frame_id;
					poseMsg.header.stamp = msg->header.stamp;
					pose_pub.publish(poseMsg);

					geometry_msgs::TransformStamped transformMsg;
					tf::transformStampedTFToMsg(stampedTransform, transformMsg);
					transform_pub.publish(transformMsg);

					geometry_msgs::Vector3Stamped positionMsg;
					positionMsg.header = transformMsg.header;
					positionMsg.vector = transformMsg.transform.translation;
					position_pub.publish(positionMsg);
				}
				//for each marker, draw info and its boundaries in the image
				for(size_t i=0; draw_markers && i < markers.size(); ++i)
				{
					markers[i].draw(resultImg,cv::Scalar(0,0,255),2);
				}


				if(camParam.isValid() && marker_size != -1)
				{
					//draw a 3d cube in each marker if there is 3d info
					for(size_t i=0; i<markers.size(); ++i)
					{
						if (draw_markers_cube) CvDrawingUtils::draw3dCube(resultImg, markers[i], camParam);
						if (draw_markers_axis) CvDrawingUtils::draw3dAxis(resultImg, markers[i], camParam);
					}
					//draw board axis
					if (probDetect > 0.0) CvDrawingUtils::draw3dAxis(resultImg, the_board_detected, camParam);
				}

				if(image_pub.getNumSubscribers() > 0)
				{
					//show input with augmented information
					cv_bridge::CvImage out_msg;
					out_msg.header.frame_id = msg->header.frame_id;
					out_msg.header.stamp = msg->header.stamp;
					out_msg.encoding = sensor_msgs::image_encodings::RGB8;
					out_msg.image = resultImg;
					image_pub.publish(out_msg.toImageMsg());
				}

				if(debug_pub.getNumSubscribers() > 0)
				{
					//show also the internal image resulting from the threshold operation
					cv_bridge::CvImage debug_msg;
					debug_msg.header.frame_id = msg->header.frame_id;
					debug_msg.header.stamp = msg->header.stamp;
					debug_msg.encoding = sensor_msgs::image_encodings::MONO8;
					debug_msg.image = mDetector.getThresholdedImage();
					debug_pub.publish(debug_msg.toImageMsg());
				}
			}
			catch (cv_bridge::Exception& e)
			{
				ROS_ERROR("cv_bridge exception: %s", e.what());
				return;
			}
		}

		// wait for one camerainfo, then shut down that subscriber
		void cam_info_callback(const sensor_msgs::CameraInfo &msg)
		{
			camParam = ar_sys::getCamParams(msg, useRectifiedImages);
			cam_info_received = true;
			cam_info_sub.shutdown();
		}
};


int main(int argc,char **argv)
{
	ros::init(argc, argv, "ar_single_board");

	ArSysSingleBoard node;

	ros::spin();
}


src/system_viewer.cpp
/**
* @file system_viewer.cpp
* @author Hamdi Sahloul
* @date September 2014
* @version 0.1
* @brief Display all the detected boards and all the cameras in a single 3D coordinates system.
*/

#include <opencv2/core/core.hpp>
#include <aruco/aruco.h>

#include <ros/ros.h>
#include <tf/transform_broadcaster.h>
#include <visualization_msgs/Marker.h>

#define VISUAL_MARKER_SIZE	0.08

enum position_t
{
	FIXED,
	RELATIVE,
	FLOATING,

	POSITION_NR_ITEMS
};

struct board_t
{
	int uid;
	std::string name;
	position_t type;

	std::string relativeName;
	tf::Transform transform;
};

struct relative_board_t
{
	board_t board;
	std::map<std::string, tf::Transform> transforms_map;
};

struct camera_t
{
	int uid;
	std::string name;
	position_t type;

	tf::Transform transform;
};

class ArSysViewer
{
	private:
		std::string map_path;
		std::map<std::string, board_t> boards_map;
		std::map<std::string, relative_board_t> relativeBoards_map;
		std::map<std::string, camera_t> cameras_map;
		double digital_filter_change_rate;

		ros::Subscriber transform_sub;
		ros::Publisher rviz_marker_pub;
		tf::TransformBroadcaster broadcaster;
		ros::NodeHandle nh;

	public:
		ArSysViewer()
			: nh("~")
		{
			transform_sub = nh.subscribe ("/transform", 1, &ArSysViewer::transform_callback, this);

			rviz_marker_pub = nh.advertise <visualization_msgs::Marker> ("visualization_marker", 0);

			nh.param<std::string>("map_path", map_path, "map.yml");
			nh.param<double>("digital_filter_change_rate", digital_filter_change_rate, 0.5);

			readFromFile(map_path.c_str());
			ROS_INFO("ArSys node started with boards map: %s", map_path.c_str());
		}

		void readFromFile ( string sfile ) throw ( cv::Exception )
		{
			try
			{
				cv::FileStorage fs ( sfile,cv::FileStorage::READ );
				readFromFile ( fs );
			}
			catch (std::exception &ex)
			{
				throw	cv::Exception ( 81818,"ArSysViewer::readFromFile",ex.what()+string(" file=)")+sfile ,__FILE__,__LINE__ );
			}
		}


		void readFromFile ( cv::FileStorage &fs ) throw ( cv::Exception )
		{
			//look for the ar_map_boards and ar_map_cameras
			if (fs["ar_map_boards"].name() != "ar_map_boards" || fs["ar_map_cameras"].name() != "ar_map_cameras")
				throw cv::Exception ( 81818,"ArSysViewer::readFromFile","invalid file type" ,__FILE__,__LINE__ );

			cv::FileNode FnBoards=fs["ar_map_boards"];
			for (cv::FileNodeIterator it = FnBoards.begin(); it != FnBoards.end(); ++it)
			{
				board_t board;

				board.uid = boards_map.size();
				board.name = (std::string)(*it)["name"];
				if ((std::string)(*it)["type"] == "fixed")
				{
					board.type = FIXED;

				}
				else if ((std::string)(*it)["type"] == "relative")
				{
					board.type = RELATIVE;
					board.relativeName = (std::string)(*it)["relativeName"];
				}
				else
				{
					if ((std::string)(*it)["type"] != "floating") ROS_WARN("Unknow board type: %s for board %s, assuming floating",
						((std::string)(*it)["type"]).c_str(),
						board.name.c_str());
					board.type = FLOATING;
					board.transform = tf::Transform::getIdentity(); //Initial position
				}
				if (board.type == FIXED || board.type == RELATIVE)
				{
					vector<float> coordinates3d;
					cv::FileNode FnPose = (*it)["pose"];
					cv::FileNodeIterator itc = FnPose.begin();

					(*itc)>>coordinates3d;
					if (coordinates3d.size() != 3)
				    		throw cv::Exception ( 81818,"ArSysViewer::readFromFile","invalid file type 3" ,__FILE__,__LINE__ );
					tf::Vector3 ref_origin(coordinates3d[0], coordinates3d[1], coordinates3d[2]);

					(*++itc)>>coordinates3d;
					if (coordinates3d.size() != 3)
				    		throw cv::Exception ( 81818,"ArSysViewer::readFromFile","invalid file type 3" ,__FILE__,__LINE__ );
					tf::Quaternion ref_rotation = tf::createQuaternionFromRPY(coordinates3d[0], coordinates3d[1], coordinates3d[2]);

					board.transform = tf::Transform(ref_rotation, ref_origin);

					if (board.type == RELATIVE)
					{
						if (relativeBoards_map.count(board.relativeName) == 0)
						{
							relative_board_t relativeBoard;

							relativeBoard.board.uid = relativeBoards_map.size();
							relativeBoard.board.name = board.relativeName;
							relativeBoard.board.type = FLOATING;
							relativeBoard.board.transform = tf::Transform::getIdentity(); //Initial position

							relativeBoards_map.insert(std::make_pair(relativeBoard.board.name, relativeBoard));
						}
						relativeBoards_map[board.relativeName].transforms_map.insert(std::make_pair(board.name, board.transform));
						board.transform = tf::Transform::getIdentity(); //Done with relative, Initial position
					}
				}

				boards_map.insert(std::make_pair(board.name, board));
			}

			cv::FileNode FnCameras=fs["ar_map_cameras"];
			for (cv::FileNodeIterator it = FnCameras.begin(); it != FnCameras.end(); ++it)
			{
				camera_t camera;

				camera.uid = cameras_map.size();
				camera.name = (std::string)(*it)["name"];
				if ((std::string)(*it)["type"] == "fixed")
				{
					camera.type = FIXED;

					vector<float> coordinates3d;
					cv::FileNode FnPose = (*it)["pose"];
					cv::FileNodeIterator itc = FnPose.begin();

					(*itc)>>coordinates3d;
					if (coordinates3d.size() != 3)
				    		throw cv::Exception ( 81818,"ArSysViewer::readFromFile","invalid file type 3" ,__FILE__,__LINE__ );
					tf::Vector3 ref_origin(coordinates3d[0], coordinates3d[1], coordinates3d[2]);

					(*++itc)>>coordinates3d;
					if (coordinates3d.size() != 3)
				    		throw cv::Exception ( 81818,"ArSysViewer::readFromFile","invalid file type 3" ,__FILE__,__LINE__ );
					tf::Quaternion ref_rotation = tf::createQuaternionFromRPY(coordinates3d[0], coordinates3d[1], coordinates3d[2]);

					camera.transform = tf::Transform(ref_rotation, ref_origin);
				}
				else
				{
					if ((std::string)(*it)["type"] != "floating") ROS_WARN("Unknow camera type: %s for camera %s, assuming floating",
						((std::string)(*it)["type"]).c_str(),
						camera.name.c_str());
					camera.type = FLOATING;
					camera.transform = tf::Transform::getIdentity(); //Initial position
				}

				cameras_map.insert(std::make_pair(camera.name, camera));
			}
		}

		tf::Transform& digital_filter(tf::Transform &dst, const tf::Transform &src)
		{
			//dst = src;

			tf::Vector3 posOld = dst.getOrigin();
			tf::Vector3 posNew = src.getOrigin();
			tf::Vector3 pos
			(
				(1 - digital_filter_change_rate) * posOld.x() + digital_filter_change_rate * posNew.x(),
				(1 - digital_filter_change_rate) * posOld.y() + digital_filter_change_rate * posNew.y(),
				(1 - digital_filter_change_rate) * posOld.z() + digital_filter_change_rate * posNew.z()
			);
			dst.setOrigin(pos);

			tf::Quaternion ornOld = dst.getRotation();
			tf::Quaternion ornNew = src.getRotation();
			tf::Quaternion orn
			(
				(1 - digital_filter_change_rate) * ornOld.x() + digital_filter_change_rate * ornNew.x(),
				(1 - digital_filter_change_rate) * ornOld.y() + digital_filter_change_rate * ornNew.y(),
				(1 - digital_filter_change_rate) * ornOld.z() + digital_filter_change_rate * ornNew.z(),
				(1 - digital_filter_change_rate) * ornOld.w() + digital_filter_change_rate * ornNew.w()
			);
			dst.setRotation(orn);

			return dst;
		}

		void transform_callback (const geometry_msgs::TransformStamped& transformMsg)
		{
			if (cameras_map.count(transformMsg.header.frame_id) == 0)
			{
				ROS_WARN("Map file does not contain an entry for the '%s' camera! Adding as floating", transformMsg.header.frame_id.c_str());

				camera_t camera;
				camera.uid = cameras_map.size();
				camera.name = transformMsg.header.frame_id;
				camera.type = FLOATING;
				camera.transform = tf::Transform::getIdentity(); //Initial position
				cameras_map.insert(std::make_pair(camera.name, camera));
			}
			if (boards_map.count(transformMsg.child_frame_id) == 0)
			{
				ROS_WARN("Map file does not contain an entry for the '%s' board! Adding as floating", transformMsg.child_frame_id.c_str());

				board_t board;
				board.uid = boards_map.size();
				board.name = transformMsg.child_frame_id;
				board.type = FLOATING;
				board.transform = tf::Transform::getIdentity(); //Initial position
				boards_map.insert(std::make_pair(board.name, board));
			}

			tf::StampedTransform stampedTransform;
			tf::transformStampedMsgToTF(transformMsg, stampedTransform);
			if (cameras_map[transformMsg.header.frame_id].type == FLOATING && boards_map[transformMsg.child_frame_id].type == FIXED)
			{
				digital_filter(cameras_map[transformMsg.header.frame_id].transform, boards_map[transformMsg.child_frame_id].transform * stampedTransform.inverse());
			}
			tf::StampedTransform camStampedTransform (cameras_map[transformMsg.header.frame_id].transform, transformMsg.header.stamp, "world", transformMsg.header.frame_id);
			broadcaster.sendTransform(camStampedTransform);

			stampedTransform.setData(boards_map[transformMsg.child_frame_id].type == FIXED ?
				boards_map[transformMsg.child_frame_id].transform :
				digital_filter(boards_map[transformMsg.child_frame_id].transform, cameras_map[transformMsg.header.frame_id].transform * stampedTransform));
			stampedTransform.frame_id_ = camStampedTransform.frame_id_;
			broadcaster.sendTransform(stampedTransform);
			if (boards_map[transformMsg.child_frame_id].type == RELATIVE)
			{
				std::string relativeName = boards_map[transformMsg.child_frame_id].relativeName;

				digital_filter(relativeBoards_map[relativeName].board.transform, boards_map[transformMsg.child_frame_id].transform * relativeBoards_map[relativeName].transforms_map[transformMsg.child_frame_id]);
				tf::StampedTransform relativeStampedTransform(relativeBoards_map[relativeName].board.transform, transformMsg.header.stamp, "world", relativeName);
				broadcaster.sendTransform(relativeStampedTransform);
			}

			//Generate visual marker
			tf::Vector3 markerOrigin (0, 0, 0.1 * VISUAL_MARKER_SIZE);
			tf::Transform m (tf::Quaternion::getIdentity(), markerOrigin);
			tf::Transform markerPose = boards_map[transformMsg.child_frame_id].transform * m;

			visualization_msgs::Marker visualMarker;
			tf::poseTFToMsg (markerPose, visualMarker.pose);

			visualMarker.header.frame_id = stampedTransform.frame_id_;
			visualMarker.header.stamp = transformMsg.header.stamp;
			visualMarker.id = boards_map[transformMsg.child_frame_id].uid;

			visualMarker.scale.x = 1.0 * VISUAL_MARKER_SIZE;
			visualMarker.scale.y = 1.0 * VISUAL_MARKER_SIZE;
			visualMarker.scale.z = 0.1 * VISUAL_MARKER_SIZE;
			visualMarker.ns = "basic_shapes";
			visualMarker.type = visualization_msgs::Marker::CUBE;
			visualMarker.action = visualization_msgs::Marker::ADD;
			visualMarker.color.r = (float)visualMarker.id / (boards_map.size() - 1);
			visualMarker.color.g = 1.0 - visualMarker.color.r;
			visualMarker.color.b = (float)boards_map[transformMsg.child_frame_id].type / (POSITION_NR_ITEMS - 1);
			visualMarker.color.a = 1.0;
			visualMarker.lifetime = ros::Duration (1.0);
			//Publish visual marker
			rviz_marker_pub.publish(visualMarker);
		}
};


int main(int argc,char **argv)
{
	ros::init(argc, argv, "ar_system_viewer");

	ArSysViewer node;

	ros::spin();
}


src/multi_boards.cpp
/**
* @file multi_boards.cpp
* @author Hamdi Sahloul
* @date September 2014
* @version 0.1
* @brief Detect multi-boards simultaneously.
*/

#include <iostream>
#include <fstream>
#include <sstream>
#include <aruco/aruco.h>
#include <aruco/boarddetector.h>
#include <aruco/cvdrawingutils.h>

#include <opencv2/core/core.hpp>
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/image_encodings.h>
#include <ar_sys/utils.h>
#include <tf/transform_broadcaster.h>
#include <tf/transform_listener.h>

using namespace aruco;

struct board_t
{
	int uid;
	std::string name;
	BoardConfiguration config;
	double marker_size;
};

class ArSysMultiBoards
{
	private:
		cv::Mat inImage, resultImg;
		aruco::CameraParameters camParam;
		bool useRectifiedImages;
		bool draw_markers;
		bool draw_markers_cube;
		bool draw_markers_axis;
		MarkerDetector mDetector;
		vector<Marker> markers;
		BoardDetector the_board_detector;
		ros::Subscriber cam_info_sub;
		bool cam_info_received;
		image_transport::Publisher image_pub;
		image_transport::Publisher debug_pub;
		ros::Publisher pose_pub;
		ros::Publisher transform_pub; 
		ros::Publisher position_pub;
		std::string boards_config;
		std::string boards_directory;
		vector<board_t> boards;

		ros::NodeHandle nh;
		image_transport::ImageTransport it;
		image_transport::Subscriber image_sub;

		tf::TransformListener _tfListener;

	public:
		ArSysMultiBoards()
			: cam_info_received(false),
			nh("~"),
			it(nh)
		{
			image_sub = it.subscribe("/image", 1, &ArSysMultiBoards::image_callback, this);
			cam_info_sub = nh.subscribe("/camera_info", 1, &ArSysMultiBoards::cam_info_callback, this);

			image_pub = it.advertise("result", 1);
			debug_pub = it.advertise("debug", 1);
			pose_pub = nh.advertise<geometry_msgs::PoseStamped>("pose", 100);
			transform_pub = nh.advertise<geometry_msgs::TransformStamped>("transform", 100);
			position_pub = nh.advertise<geometry_msgs::Vector3Stamped>("position", 100);

			nh.param<std::string>("boards_config", boards_config, "boardsConfiguration.yml");
			nh.param<std::string>("boards_directory", boards_directory, "./data");
			nh.param<bool>("image_is_rectified", useRectifiedImages, true);
			nh.param<bool>("draw_markers", draw_markers, false);
			nh.param<bool>("draw_markers_cube", draw_markers_cube, false);
			nh.param<bool>("draw_markers_axis", draw_markers_axis, false);

			readFromFile(boards_config.c_str());
			ROS_INFO("ArSys node started with boards configuration: %s", boards_config.c_str());
		}

		void readFromFile ( string sfile ) throw ( cv::Exception )
		{
			try
			{
				cv::FileStorage fs ( sfile,cv::FileStorage::READ );
				readFromFile ( fs );
			}
			catch (std::exception &ex)
			{
				throw	cv::Exception ( 81818,"ArSysMultiBoards::readFromFile",ex.what()+string(" file=)")+sfile ,__FILE__,__LINE__ );
			}
		}


		void readFromFile ( cv::FileStorage &fs ) throw ( cv::Exception )
		{
			//look for the ar_sys_boards
			if (fs["ar_sys_boards"].name() != "ar_sys_boards")
				throw cv::Exception ( 81818,"ArSysMultiBoards::readFromFile","invalid file type" ,__FILE__,__LINE__ );

			cv::FileNode FnBoards=fs["ar_sys_boards"];
			for (cv::FileNodeIterator it = FnBoards.begin(); it != FnBoards.end(); ++it)
			{
				board_t board;

				board.uid = boards.size();
				board.name = (std::string)(*it)["name"];
				board.marker_size = (double)(*it)["marker_size"];

				std::string path(boards_directory);
				path.append("/");
				path.append((std::string)(*it)["path"]);
				board.config.readFromFile(path);

				boards.push_back(board);
			}

			ROS_ASSERT(boards.size() > 0);
		}

		void image_callback(const sensor_msgs::ImageConstPtr& msg)
		{
			if(!cam_info_received) return;

			cv_bridge::CvImagePtr cv_ptr;
			try
			{
				cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::RGB8);
				inImage = cv_ptr->image;
				resultImg = cv_ptr->image.clone();

				//detection results will go into "markers"
				markers.clear();

				//Ok, let's detect
				double min_size = boards[0].marker_size;
				for (int board_index = 1; board_index < boards.size(); board_index++)
					if (min_size > boards[board_index].marker_size) min_size = boards[board_index].marker_size;
				mDetector.detect(inImage, markers, camParam, min_size, false);

				for (int board_index = 0; board_index < boards.size(); board_index++)
				{
					Board board_detected;

					//Detection of the board
					float probDetect = the_board_detector.detect(markers, boards[board_index].config, board_detected, camParam, boards[board_index].marker_size);
					if (probDetect > 0.0)
					{
						tf::Transform transform = ar_sys::getTf(board_detected.Rvec, board_detected.Tvec);

						tf::StampedTransform stampedTransform(transform, msg->header.stamp, msg->header.frame_id, boards[board_index].name);

						geometry_msgs::PoseStamped poseMsg;
						tf::poseTFToMsg(transform, poseMsg.pose);
						poseMsg.header.frame_id = msg->header.frame_id;
						poseMsg.header.stamp = msg->header.stamp;
						pose_pub.publish(poseMsg);

						geometry_msgs::TransformStamped transformMsg;
						tf::transformStampedTFToMsg(stampedTransform, transformMsg);
						transform_pub.publish(transformMsg);

						geometry_msgs::Vector3Stamped positionMsg;
						positionMsg.header = transformMsg.header;
						positionMsg.vector = transformMsg.transform.translation;
						position_pub.publish(positionMsg);

						if(camParam.isValid())
						{
							//draw board axis
							CvDrawingUtils::draw3dAxis(resultImg, board_detected, camParam);
						}
					}
				}

				//for each marker, draw info and its boundaries in the image
				for(size_t i=0; draw_markers && i < markers.size(); ++i)
				{
					markers[i].draw(resultImg,cv::Scalar(0,0,255),2);
				}

				if(camParam.isValid())
				{
					//draw a 3d cube in each marker if there is 3d info
					for(size_t i=0; i<markers.size(); ++i)
					{
						if (draw_markers_cube) CvDrawingUtils::draw3dCube(resultImg, markers[i], camParam);
						if (draw_markers_axis) CvDrawingUtils::draw3dAxis(resultImg, markers[i], camParam);
					}
				}

				if(image_pub.getNumSubscribers() > 0)
				{
					//show input with augmented information
					cv_bridge::CvImage out_msg;
					out_msg.header.frame_id = msg->header.frame_id;
					out_msg.header.stamp = msg->header.stamp;
					out_msg.encoding = sensor_msgs::image_encodings::RGB8;
					out_msg.image = resultImg;
					image_pub.publish(out_msg.toImageMsg());
				}

				if(debug_pub.getNumSubscribers() > 0)
				{
					//show also the internal image resulting from the threshold operation
					cv_bridge::CvImage debug_msg;
					debug_msg.header.frame_id = msg->header.frame_id;
					debug_msg.header.stamp = msg->header.stamp;
					debug_msg.encoding = sensor_msgs::image_encodings::MONO8;
					debug_msg.image = mDetector.getThresholdedImage();
					debug_pub.publish(debug_msg.toImageMsg());
				}
			}
			catch (cv_bridge::Exception& e)
			{
				ROS_ERROR("cv_bridge exception: %s", e.what());
				return;
			}
		}

		// wait for one camerainfo, then shut down that subscriber
		void cam_info_callback(const sensor_msgs::CameraInfo &msg)
		{
			camParam = ar_sys::getCamParams(msg, useRectifiedImages);
			cam_info_received = true;
			cam_info_sub.shutdown();
		}
};


int main(int argc,char **argv)
{
	ros::init(argc, argv, "ar_multi_boards");

	ArSysMultiBoards node;

	ros::spin();
}
